File: testers/mariadb/java/src/java/testdb/ViewTest.java
package testdb;

import static org.junit.Assert.*;

import dev.typr.foundations.data.Uint1;
import dev.typr.foundations.data.Uint2;
import java.math.BigDecimal;
import java.util.Optional;
import org.junit.Test;
import org.mariadb.jdbc.type.Point;
import testdb.brands.*;
import testdb.customer_status.CustomerStatusId;
import testdb.customers.*;
import testdb.customtypes.Defaulted.Provided;
import testdb.customtypes.Defaulted.UseDefault;
import testdb.inventory.*;
import testdb.order_items.*;
import testdb.orders.*;
import testdb.products.*;
import testdb.reviews.*;
import testdb.userdefined.Email;
import testdb.userdefined.FirstName;
import testdb.userdefined.IsApproved;
import testdb.userdefined.LastName;
import testdb.v_customer_summary.*;
import testdb.v_product_catalog.*;
import testdb.warehouses.*;

/**
 * Tests for views - read-only operations using generated view repositories. Note: The database has
 * seeded customer_status data including 'active', 'pending', 'suspended', 'closed'. Tests use these
 * existing statuses rather than inserting new ones.
 */
public class ViewTest {
  private final VCustomerSummaryViewRepoImpl customerSummaryRepo =
      new VCustomerSummaryViewRepoImpl();
  private final VProductCatalogViewRepoImpl productCatalogRepo = new VProductCatalogViewRepoImpl();
  private final CustomersRepoImpl customersRepo = new CustomersRepoImpl();
  private final ProductsRepoImpl productsRepo = new ProductsRepoImpl();
  private final BrandsRepoImpl brandsRepo = new BrandsRepoImpl();
  private final OrdersRepoImpl ordersRepo = new OrdersRepoImpl();
  private final OrderItemsRepoImpl orderItemsRepo = new OrderItemsRepoImpl();
  private final ReviewsRepoImpl reviewsRepo = new ReviewsRepoImpl();
  private final WarehousesRepoImpl warehousesRepo = new WarehousesRepoImpl();
  private final InventoryRepoImpl inventoryRepo = new InventoryRepoImpl();

  @Test
  public void testCustomerSummaryViewSelectAll() {
    MariaDbTestHelper.run(
        c -> {
          customersRepo.insert(
              new CustomersRowUnsaved(
                  new Email("view1@example.com"),
                  "hash1".getBytes(),
                  new FirstName("View"),
                  new LastName("Customer1")),
              c);
          customersRepo.insert(
              new CustomersRowUnsaved(
                  new Email("view2@example.com"),
                  "hash2".getBytes(),
                  new FirstName("View"),
                  new LastName("Customer2")),
              c);

          var summaries = customerSummaryRepo.selectAll(c);
          assertEquals(2, summaries.size());
        });
  }

  @Test
  public void testCustomerSummaryViewFields() {
    MariaDbTestHelper.run(
        c -> {
          customersRepo.insert(
              new CustomersRowUnsaved(
                      new Email("summary@example.com"),
                      "hash".getBytes(),
                      new FirstName("Summary"),
                      new LastName("Test"))
                  .withStatus(new Provided<>(new CustomerStatusId("suspended")))
                  .withTier(new Provided<>("gold")),
              c);

          var summaries = customerSummaryRepo.selectAll(c);
          assertEquals(1, summaries.size());

          var summary = summaries.get(0);
          assertEquals(new Email("summary@example.com"), summary.email());
          assertEquals(Optional.of("Summary Test"), summary.fullName());
          assertEquals("gold", summary.tier());
          assertEquals("suspended", summary.status().value());
          assertEquals(0L, summary.totalOrders().longValue());
          assertEquals(new BigDecimal("0.0000"), summary.lifetimeValue());
        });
  }

  @Test
  public void testCustomerSummaryViewWithOrders() {
    MariaDbTestHelper.run(
        c -> {
          var customer =
              customersRepo.insert(
                  new CustomersRowUnsaved(
                      new Email("orders@example.com"),
                      "hash".getBytes(),
                      new FirstName("With"),
                      new LastName("Orders")),
                  c);

          var brand = brandsRepo.insert(new BrandsRowUnsaved("TestBrand", "test-brand"), c);
          var product =
              productsRepo.insert(
                  new ProductsRowUnsaved("SKU-ORDER", "Order Product", new BigDecimal("100.00")),
                  c);

          var order =
              ordersRepo.insert(
                  new OrdersRowUnsaved(
                      "ORD-001",
                      customer.customerId(),
                      new BigDecimal("100.00"),
                      new BigDecimal("110.00")),
                  c);

          orderItemsRepo.insert(
              new OrderItemsRowUnsaved(
                  order.orderId(),
                  product.productId(),
                  "SKU-ORDER",
                  "Order Product",
                  Uint2.of(1),
                  new BigDecimal("100.00"),
                  new BigDecimal("100.00"),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>()),
              c);

          var summaries = customerSummaryRepo.selectAll(c);
          assertEquals(1, summaries.size());
          assertTrue(summaries.get(0).totalOrders() >= 1);
        });
  }

  @Test
  public void testProductCatalogViewSelectAll() {
    MariaDbTestHelper.run(
        c -> {
          var brand = brandsRepo.insert(new BrandsRowUnsaved("CatalogBrand", "catalog-brand"), c);

          productsRepo.insert(
              new ProductsRowUnsaved(
                  "CAT-SKU-001",
                  "Catalog Product 1",
                  new BigDecimal("25.00"),
                  new Provided<>(Optional.of(brand.brandId())),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new Provided<>("active"),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>()),
              c);
          productsRepo.insert(
              new ProductsRowUnsaved(
                  "CAT-SKU-002",
                  "Catalog Product 2",
                  new BigDecimal("50.00"),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new Provided<>("active"),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>()),
              c);

          var catalog = productCatalogRepo.selectAll(c);
          assertEquals(2, catalog.size());
        });
  }

  @Test
  public void testProductCatalogViewFields() {
    MariaDbTestHelper.run(
        c -> {
          var brand = brandsRepo.insert(new BrandsRowUnsaved("FieldsBrand", "fields-brand"), c);

          productsRepo.insert(
              new ProductsRowUnsaved(
                  "FIELDS-SKU",
                  "Fields Product",
                  new BigDecimal("99.99"),
                  new Provided<>(Optional.of(brand.brandId())),
                  new Provided<>(Optional.of("Short description")),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new Provided<>("active"),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>()),
              c);

          var catalog = productCatalogRepo.selectAll(c);
          assertEquals(1, catalog.size());

          var row = catalog.get(0);
          assertEquals("FIELDS-SKU", row.sku());
          assertEquals("Fields Product", row.name());
          assertEquals(0, new BigDecimal("99.99").compareTo(row.basePrice()));
          assertEquals("active", row.status());
          assertEquals(Optional.of("FieldsBrand"), row.brandName());
          assertEquals(Optional.of("Short description"), row.shortDescription());
          assertEquals(0L, row.reviewCount().longValue());
        });
  }

  @Test
  public void testProductCatalogViewWithInventory() {
    MariaDbTestHelper.run(
        c -> {
          var warehouse =
              warehousesRepo.insert(
                  new WarehousesRowUnsaved(
                      "WH001", "Main Warehouse", "123 Warehouse St", new Point(0.0, 0.0)),
                  c);

          var product =
              productsRepo.insert(
                  new ProductsRowUnsaved(
                      "INV-SKU",
                      "Inventory Product",
                      new BigDecimal("75.00"),
                      new UseDefault<>(),
                      new UseDefault<>(),
                      new UseDefault<>(),
                      new UseDefault<>(),
                      new UseDefault<>(),
                      new UseDefault<>(),
                      new Provided<>("active"),
                      new UseDefault<>(),
                      new UseDefault<>(),
                      new UseDefault<>(),
                      new UseDefault<>(),
                      new UseDefault<>(),
                      new UseDefault<>(),
                      new UseDefault<>()),
                  c);

          inventoryRepo.insert(
              new InventoryRowUnsaved(
                  product.productId(),
                  warehouse.warehouseId(),
                  new Provided<>(100),
                  new Provided<>(10),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>()),
              c);

          var catalog = productCatalogRepo.selectAll(c);
          assertEquals(1, catalog.size());
          assertEquals(new BigDecimal("90"), catalog.get(0).availableQuantity());
        });
  }

  @Test
  public void testProductCatalogViewWithReviews() {
    MariaDbTestHelper.run(
        c -> {
          var customer =
              customersRepo.insert(
                  new CustomersRowUnsaved(
                      new Email("reviewer@example.com"),
                      "hash".getBytes(),
                      new FirstName("Reviewer"),
                      new LastName("User")),
                  c);

          var product =
              productsRepo.insert(
                  new ProductsRowUnsaved(
                      "REVIEW-SKU",
                      "Reviewed Product",
                      new BigDecimal("150.00"),
                      new UseDefault<>(),
                      new UseDefault<>(),
                      new UseDefault<>(),
                      new UseDefault<>(),
                      new UseDefault<>(),
                      new UseDefault<>(),
                      new Provided<>("active"),
                      new UseDefault<>(),
                      new UseDefault<>(),
                      new UseDefault<>(),
                      new UseDefault<>(),
                      new UseDefault<>(),
                      new UseDefault<>(),
                      new UseDefault<>()),
                  c);

          reviewsRepo.insert(
              new ReviewsRowUnsaved(
                  product.productId(),
                  customer.customerId(),
                  Uint1.of(5),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new Provided<>(new IsApproved(true)),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>()),
              c);
          reviewsRepo.insert(
              new ReviewsRowUnsaved(
                  product.productId(),
                  customer.customerId(),
                  Uint1.of(4),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new Provided<>(new IsApproved(true)),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>()),
              c);

          var catalog = productCatalogRepo.selectAll(c);
          assertEquals(1, catalog.size());
          assertEquals(2L, catalog.get(0).reviewCount().longValue());
          assertTrue(catalog.get(0).avgRating().compareTo(new BigDecimal("4")) >= 0);
        });
  }

  @Test
  public void testViewDSLSelect() {
    MariaDbTestHelper.run(
        c -> {
          customersRepo.insert(
              new CustomersRowUnsaved(
                      new Email("dsl1@example.com"),
                      "hash1".getBytes(),
                      new FirstName("DSL"),
                      new LastName("Bronze"))
                  .withTier(new Provided<>("bronze")),
              c);
          customersRepo.insert(
              new CustomersRowUnsaved(
                      new Email("dsl2@example.com"),
                      "hash2".getBytes(),
                      new FirstName("DSL"),
                      new LastName("Gold"))
                  .withTier(new Provided<>("gold")),
              c);
          customersRepo.insert(
              new CustomersRowUnsaved(
                      new Email("dsl3@example.com"),
                      "hash3".getBytes(),
                      new FirstName("DSL"),
                      new LastName("Gold2"))
                  .withTier(new Provided<>("gold")),
              c);

          var goldCustomers =
              customerSummaryRepo.select().where(f -> f.tier().isEqual("gold")).toList(c);
          assertEquals(2, goldCustomers.size());

          var specificCustomer =
              customerSummaryRepo
                  .select()
                  .where(f -> f.email().isEqual(new Email("dsl1@example.com")))
                  .toList(c);
          assertEquals(1, specificCustomer.size());
          assertEquals("bronze", specificCustomer.get(0).tier());
        });
  }

  @Test
  public void testProductCatalogViewDSLSelect() {
    MariaDbTestHelper.run(
        c -> {
          productsRepo.insert(
              new ProductsRowUnsaved(
                  "DSL-PROD-1",
                  "Cheap Product",
                  new BigDecimal("10.00"),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new Provided<>("active"),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>()),
              c);
          productsRepo.insert(
              new ProductsRowUnsaved(
                  "DSL-PROD-2",
                  "Expensive Product",
                  new BigDecimal("1000.00"),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new Provided<>("active"),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>(),
                  new UseDefault<>()),
              c);
          productsRepo.insert(
              new ProductsRowUnsaved("DSL-PROD-3", "Draft Product", new BigDecimal("50.00")), c);

          var activeProducts =
              productCatalogRepo.select().where(f -> f.status().isEqual("active")).toList(c);
          assertEquals(2, activeProducts.size());

          var specificProduct =
              productCatalogRepo.select().where(f -> f.sku().isEqual("DSL-PROD-1")).toList(c);
          assertEquals(1, specificProduct.size());
          assertEquals("Cheap Product", specificProduct.get(0).name());
        });
  }
}

File: testers/mariadb/scala/src/scala/testdb/ViewTest.scala
package testdb

import dev.typr.foundations.data.{Uint1, Uint2}
import org.mariadb.jdbc.`type`.Point
import org.scalatest.funsuite.AnyFunSuite
import testdb.brands.*
import testdb.customer_status.CustomerStatusId
import testdb.customers.*
import testdb.customtypes.Defaulted.Provided
import testdb.inventory.*
import testdb.order_items.*
import testdb.orders.*
import testdb.products.*
import testdb.reviews.*
import testdb.v_customer_summary.*
import testdb.v_product_catalog.*
import testdb.warehouses.*

/** Tests for views - read-only operations using generated view repositories. Note: The database has seeded customer_status data including 'active', 'pending', 'suspended', 'closed'. Tests use these
  * existing statuses rather than inserting new ones.
  */
class ViewTest extends AnyFunSuite {
  val customerSummaryRepo: VCustomerSummaryViewRepoImpl = new VCustomerSummaryViewRepoImpl
  val productCatalogRepo: VProductCatalogViewRepoImpl = new VProductCatalogViewRepoImpl
  val customersRepo: CustomersRepoImpl = new CustomersRepoImpl
  val productsRepo: ProductsRepoImpl = new ProductsRepoImpl
  val brandsRepo: BrandsRepoImpl = new BrandsRepoImpl
  val ordersRepo: OrdersRepoImpl = new OrdersRepoImpl
  val orderItemsRepo: OrderItemsRepoImpl = new OrderItemsRepoImpl
  val reviewsRepo: ReviewsRepoImpl = new ReviewsRepoImpl
  val warehousesRepo: WarehousesRepoImpl = new WarehousesRepoImpl
  val inventoryRepo: InventoryRepoImpl = new InventoryRepoImpl

  test("customerSummaryViewSelectAll") {
    withConnection { c =>
      given java.sql.Connection = c

      customersRepo.insert(
        CustomersRowUnsaved(
          email = "view1@example.com",
          passwordHash = "hash1".getBytes,
          firstName = "View",
          lastName = "Customer1"
        )
      )
      customersRepo.insert(
        CustomersRowUnsaved(
          email = "view2@example.com",
          passwordHash = "hash2".getBytes,
          firstName = "View",
          lastName = "Customer2"
        )
      )

      val summaries = customerSummaryRepo.selectAll
      assert(summaries.size == 2)
    }
  }

  test("customerSummaryViewFields") {
    withConnection { c =>
      given java.sql.Connection = c

      customersRepo.insert(
        CustomersRowUnsaved(
          email = "summary@example.com",
          passwordHash = "hash".getBytes,
          firstName = "Summary",
          lastName = "Test",
          status = Provided(CustomerStatusId("suspended")),
          tier = Provided("gold")
        )
      )

      val summaries = customerSummaryRepo.selectAll
      val _ = assert(summaries.size == 1)

      val summary = summaries.head
      val _ = assert(summary.email == "summary@example.com")
      val _ = assert(summary.fullName == Some("Summary Test"))
      val _ = assert(summary.tier == "gold")
      val _ = assert(summary.status.value == "suspended")
      val _ = assert(summary.totalOrders == 0L)
      assert(summary.lifetimeValue == BigDecimal("0.0000"))
    }
  }

  test("customerSummaryViewWithOrders") {
    withConnection { c =>
      given java.sql.Connection = c

      val customer = customersRepo.insert(
        CustomersRowUnsaved(
          email = "orders@example.com",
          passwordHash = "hash".getBytes,
          firstName = "With",
          lastName = "Orders"
        )
      )

      val brand = brandsRepo.insert(BrandsRowUnsaved(name = "TestBrand", slug = "test-brand"))
      val product = productsRepo.insert(
        ProductsRowUnsaved(
          sku = "SKU-ORDER",
          name = "Order Product",
          basePrice = BigDecimal("100.00")
        )
      )

      val order = ordersRepo.insert(
        OrdersRowUnsaved(
          orderNumber = "ORD-001",
          customerId = customer.customerId,
          subtotal = BigDecimal("100.00"),
          totalAmount = BigDecimal("110.00")
        )
      )

      orderItemsRepo.insert(
        OrderItemsRowUnsaved(
          orderId = order.orderId,
          productId = product.productId,
          sku = "SKU-ORDER",
          productName = "Order Product",
          quantity = Uint2.of(1),
          unitPrice = BigDecimal("100.00"),
          lineTotal = BigDecimal("100.00")
        )
      )

      val summaries = customerSummaryRepo.selectAll
      val _ = assert(summaries.size == 1)
      assert(summaries.head.totalOrders >= 1)
    }
  }

  test("productCatalogViewSelectAll") {
    withConnection { c =>
      given java.sql.Connection = c

      val brand = brandsRepo.insert(BrandsRowUnsaved(name = "CatalogBrand", slug = "catalog-brand"))

      productsRepo.insert(
        ProductsRowUnsaved(
          sku = "CAT-SKU-001",
          name = "Catalog Product 1",
          basePrice = BigDecimal("25.00"),
          brandId = Provided(Some(brand.brandId)),
          status = Provided("active")
        )
      )
      productsRepo.insert(
        ProductsRowUnsaved(
          sku = "CAT-SKU-002",
          name = "Catalog Product 2",
          basePrice = BigDecimal("50.00"),
          status = Provided("active")
        )
      )

      val catalog = productCatalogRepo.selectAll
      assert(catalog.size == 2)
    }
  }

  test("productCatalogViewFields") {
    withConnection { c =>
      given java.sql.Connection = c

      val brand = brandsRepo.insert(BrandsRowUnsaved(name = "FieldsBrand", slug = "fields-brand"))

      productsRepo.insert(
        ProductsRowUnsaved(
          sku = "FIELDS-SKU",
          name = "Fields Product",
          basePrice = BigDecimal("99.99"),
          brandId = Provided(Some(brand.brandId)),
          shortDescription = Provided(Some("Short description")),
          status = Provided("active")
        )
      )

      val catalog = productCatalogRepo.selectAll
      val _ = assert(catalog.size == 1)

      val row = catalog.head
      val _ = assert(row.sku == "FIELDS-SKU")
      val _ = assert(row.name == "Fields Product")
      val _ = assert(row.basePrice.compareTo(BigDecimal("99.99")) == 0)
      val _ = assert(row.status == "active")
      val _ = assert(row.brandName == Some("FieldsBrand"))
      val _ = assert(row.shortDescription == Some("Short description"))
      assert(row.reviewCount == 0L)
    }
  }

  test("productCatalogViewWithInventory") {
    withConnection { c =>
      given java.sql.Connection = c

      val warehouse = warehousesRepo.insert(
        WarehousesRowUnsaved(
          code = "WH001",
          name = "Main Warehouse",
          address = "123 Warehouse St",
          location = new Point(0.0, 0.0)
        )
      )

      val product = productsRepo.insert(
        ProductsRowUnsaved(
          sku = "INV-SKU",
          name = "Inventory Product",
          basePrice = BigDecimal("75.00"),
          status = Provided("active")
        )
      )

      inventoryRepo.insert(
        InventoryRowUnsaved(
          productId = product.productId,
          warehouseId = warehouse.warehouseId,
          quantityOnHand = Provided(100),
          quantityReserved = Provided(10)
        )
      )

      val catalog = productCatalogRepo.selectAll
      val _ = assert(catalog.size == 1)
      assert(catalog.head.availableQuantity == BigDecimal("90"))
    }
  }

  test("productCatalogViewWithReviews") {
    withConnection { c =>
      given java.sql.Connection = c

      val customer = customersRepo.insert(
        CustomersRowUnsaved(
          email = "reviewer@example.com",
          passwordHash = "hash".getBytes,
          firstName = "Reviewer",
          lastName = "User"
        )
      )

      val product = productsRepo.insert(
        ProductsRowUnsaved(
          sku = "REVIEW-SKU",
          name = "Reviewed Product",
          basePrice = BigDecimal("150.00"),
          status = Provided("active")
        )
      )

      reviewsRepo.insert(
        ReviewsRowUnsaved(
          productId = product.productId,
          customerId = customer.customerId,
          rating = Uint1.of(5),
          isApproved = Provided(true)
        )
      )
      reviewsRepo.insert(
        ReviewsRowUnsaved(
          productId = product.productId,
          customerId = customer.customerId,
          rating = Uint1.of(4),
          isApproved = Provided(true)
        )
      )

      val catalog = productCatalogRepo.selectAll
      val _ = assert(catalog.size == 1)
      val _ = assert(catalog.head.reviewCount == 2L)
      assert(catalog.head.avgRating.compareTo(BigDecimal("4")) >= 0)
    }
  }

  test("viewDSLSelect") {
    withConnection { c =>
      given java.sql.Connection = c

      customersRepo.insert(
        CustomersRowUnsaved(
          email = "dsl1@example.com",
          passwordHash = "hash1".getBytes,
          firstName = "DSL",
          lastName = "Bronze",
          tier = Provided("bronze")
        )
      )
      customersRepo.insert(
        CustomersRowUnsaved(
          email = "dsl2@example.com",
          passwordHash = "hash2".getBytes,
          firstName = "DSL",
          lastName = "Gold",
          tier = Provided("gold")
        )
      )
      customersRepo.insert(
        CustomersRowUnsaved(
          email = "dsl3@example.com",
          passwordHash = "hash3".getBytes,
          firstName = "DSL",
          lastName = "Gold2",
          tier = Provided("gold")
        )
      )

      val goldCustomers = customerSummaryRepo.select
        .where(f => f.tier.isEqual("gold"))
        .toList
      val _ = assert(goldCustomers.size == 2)

      val specificCustomer = customerSummaryRepo.select
        .where(f => f.email.isEqual("dsl1@example.com"))
        .toList
      val _ = assert(specificCustomer.size == 1)
      assert(specificCustomer.head.tier == "bronze")
    }
  }

  test("productCatalogViewDSLSelect") {
    withConnection { c =>
      given java.sql.Connection = c

      productsRepo.insert(
        ProductsRowUnsaved(
          sku = "DSL-PROD-1",
          name = "Cheap Product",
          basePrice = BigDecimal("10.00"),
          status = Provided("active")
        )
      )
      productsRepo.insert(
        ProductsRowUnsaved(
          sku = "DSL-PROD-2",
          name = "Expensive Product",
          basePrice = BigDecimal("1000.00"),
          status = Provided("active")
        )
      )
      productsRepo.insert(
        ProductsRowUnsaved(
          sku = "DSL-PROD-3",
          name = "Draft Product",
          basePrice = BigDecimal("50.00")
        )
      )

      val activeProducts = productCatalogRepo.select
        .where(f => f.status.isEqual("active"))
        .toList
      val _ = assert(activeProducts.size == 2)

      val specificProduct = productCatalogRepo.select
        .where(f => f.sku.isEqual("DSL-PROD-1"))
        .toList
      val _ = assert(specificProduct.size == 1)
      assert(specificProduct.head.name == "Cheap Product")
    }
  }
}
